src/metabase/public_settings/premium_features.clj

```diff
@@ -240,3 +241,169 @@
+  (str/starts-with? (name (ns-name *ns*)) "metabase-enterprise"))
+
+(def ^:private valid-options #{:feature :fallback})
+
+(defn- parse-defenterprise-options
+  "Parses a list of args to defenterprise for keyword options and their corresponding values. Option keywords can be in
+  any order, but must be members of the `valid-options` set. Returns an [options-map other-args] pair."
+  [args]
+  (let [[option-name more] (u/optional valid-options args)]
+    (if-not option-name
+      [nil more]
+      (let [[option-value & other-options] more
+            [options-map other-args]      (parse-defenterprise-options other-options)]
+        [(merge options-map {option-name option-value})
+         other-args]))))
+
+(defn- parse-result-schema
+  "Detects the schema for the return value of an enterprise function, indicated with the keyword :-, and returns
+  [schema remaining-args] if it exists. Else returns [nil args]."
+  [args]
+  (let [[schema-indicator [schema & more]] (u/optional #{:-} args)]
+    (if schema-indicator
+      [schema more]
+      [nil args])))
+
+(defn- parse-defenterprise-args
+  [defenterprise-args]
+  (let [[result-schema more]    (parse-result-schema defenterprise-args)
+        [docstr more]           (u/optional string? more)
+        [ee-ns more]            (u/optional symbol? more)
+        [options [args & more]] (parse-defenterprise-options more)
+        [arg->schema body]      (u/optional (every-pred map? #(every? symbol? (keys %))) more)]
+    {:ee-ns         ee-ns
+     :docstr        docstr
+     :options       options
+     :args          args
+     :arg->schema   arg->schema
+     :result-schema result-schema
+     :body          body}))
+
+(defn validate-args
+  "Generate a series of s/validate calls for the arg and schema pairs in `arg->schema`."
+  [arg->schema]
+  (for [[arg schema] arg->schema]
+    `(s/validate ~schema ~arg)))
+
+(defonce ^{:doc "A map from EE functions (as [namespace function] tuples) to anonymous fns which have the same body as
+                their OSS equivalents. These fns are called when the EE function must fallback to the OSS behavior due
+                to absence of a feature flag."}
+  ee-registry
+  (atom {}))
+
+(defn- register-mapping!
+  "Adds mapping from EE->OSS fn to the `ee-registry`."
+  [fn-name ee-ns args body]
+  (let [ee-parts  [ee-ns fn-name]
+        oss-fn    (eval `(fn ~args ~@body))]
+    (swap! ee-registry assoc ee-parts oss-fn)))
+
+(defn missing-premium-token-exception
+  "The default exception to throw when an EE function is called, but a required premium feature is not present."
+  [fn-name feature]
+  (let [message (if (= feature :any)
+                  (trs "The {0} function requires a valid premium token"
+                       fn-name)
+                  (trs "The {0} function requires a valid premium token with the {1} feature"
+                       fn-name
+                       (name feature)))]
+    (ex-info message {:status-code 402})))
```

Comment:
If the EE function is called and we don't have the feature, shouldn't it just fall back to the OSS version (assuming one exists?)